src/art/art.c

```diff
@@ -0,0 +1,1622 @@
+#include <assert.h>
+#include <roaring/art/art.h>
+#include <roaring/memory.h>
+#include <stdio.h>
+#include <string.h>
+
+#define ART_NODE4_TYPE 0
+#define ART_NODE16_TYPE 1
+#define ART_NODE48_TYPE 2
+#define ART_NODE256_TYPE 3
+#define ART_NUM_TYPES 4
+
+// Node48 placeholder value to indicate no child is present at this key index.
+#define ART_NODE48_EMPTY_VAL 48
+
+// We use the least significant bit of node pointers to indicate whether a node
+// is a leaf or an inner node. This is never surfaced to the user.
+//
+// Using pointer tagging to indicate leaves not only saves a bit of memory by
+// sparing the typecode, but also allows us to use an intrusive leaf struct.
+// Using an intrusive leaf struct leaves leaf allocation up to the user. Upon
+// deallocation of the ART, we know not to free the leaves without having to
+// dereference the leaf pointers.
+//
+// All internal operations on leaves should use CAST_LEAF before using the leaf.
+// The only places that use SET_LEAF are locations where a field is directly
+// assigned to a leaf pointer. After using SET_LEAF, the leaf should be treated
+// as a node of unknown type.
+#define IS_LEAF(p) (((uintptr_t)p & 1))
+#define SET_LEAF(p) ((art_node_t *)((uintptr_t)p | 1))
+#define CAST_LEAF(p) ((art_leaf_t *)((void *)((uintptr_t)p & ~1)))
+
+#ifdef __cplusplus
+extern "C" {
+namespace roaring {
+namespace internal {
+#endif
+
+typedef uint8_t art_typecode_t;
+
+// Base node struct.
+typedef struct art_node_s {
```

Comment:
That's somewhat odd. It is hard to see how an empty struct would ever be useful.

Reply:
Changed to `void`.